### ISSUE TITLE
Hill climb rankings

### ISSUE THREAD

**Issue by rogernclarke** [rogernclarke/Rankings#63]
Produce a separate ranking of hill climb results

**Comment by rogernclarke** [comment]
My first thoughts are that the data processing / storage is probably the easy part.

From your side we'll need to decide what the rules are for the rankings (how many events etc.)

I think the data is best loaded into mysql in the existing tables as the data structure is the same as what we already have. We'll obviously have to be able to distinguish HCs from other TTs, so we'll need an HC indicator either on each event, each ride, or possibly both. We'll then need additional columns on the rider history table for current hc rank (male and female separate) and hc score. Could that be all we need?

The display is going to be a bite harder. We'll need a list of rankings that behaves like the current rankings but runs off the HC ranks. And we'll need to be able to display HC results on the rider page. Events page shouldn't change other than an indication it is a hill climb event. We are going to need a neat and intuitive way to switch between TT mode and HC mode.

Anything else?

**Comment by rogernclarke** [comment]
Thinking a little about how hill climbs and tts can co-exist on the website from a user point of view...

- For the list of events I think it makes sense to just include hill climbs in the list of all events, but to add a filter option to the existing distance filter to display just hill climbs.
- For the rider's profile page I imagine riders would want a single page for themselves which includes tts and hcs. The issue is that at present the list indicates which results are counting for the ranking. We would need to allow hcs and tts to be separately listed so that inclusion in the ranking can be visually indicated (maybe a switch above the list to toggle between highlighting hc counting rides and tt couting rides. Needs a bit more thought.
- For the rankings a separate menu option is needed but with the same display as for tts showing ranked riders expanding to show their counting hc rides. This can be done by having two sub-menu items under rankings, one for tts and one for hcs.
- Homepage - currently shows tt top rankings. Maybe the homepage has to be designed as part of bringing in hc rankings.

From my point of view sorting out the homepage and the filtering of rides on the rider profile page are going to be the big bits of work.

**Comment by drnickw** [comment]
Sounds good. Do you think it would be easier to display the riders HC and
TTs separately. I`m not sure there is much crossover between the two and it
would solve the highlighting problem.

**Comment by rogernclarke** [comment]
I've started to look in detail at the database changes needed.

There will be a new column on the events table called hill_climb_ind set to true (1) for a hillclimb and false (1) for a tt. This directly corresponds with an event type field on the CTT list of events. There is no need for us to also have this indicator against each ride on the rides table.
I then looked at what we need to add to the rider_history table. I was about to add a whole heap of columns to match those we store for the tt rankings, but then I realised it will nearly double the size of an already big table by storing weekly changes to rankings during the off season for hill climbs when mostly nothing changes. So I think we need a new rider_history table just for hill climbs which we should only update during the hill climb season. Could I propose a first run on the last Wednesday of August to catch all of the summer hill climbs (those events can still be scored in the week the results are posted but won't update rankings until end of August). And then the last run early November (not sure when last results will come in). Does that make sense for you?
I will then add the additional hc columns to the rider_current table so that the app only queries the current data for both tts and hcs out of the one table.

**Comment by drnickw** [comment]
Sounds good, the fewer extra bits of data the better. I’ll need to score the events thru the year as they happen cos time will change the scores; I guess we can score them on the website under HCs but not update the HC ranking till August.
I must admit I haven’t done any SD development for ages now, just housekeeping and correcting errors and omissions. I’ll need to raise my game again to get the HC processing in for the end of August but that should be possible - I’ve got two weeks away with potentially plenty of down time.
Have a look at the email conversation I’ve been having with Neil P on the “data” email. Not sure how much we want to promote the site but he makes lots of interesting suggestions.

**Comment by rogernclarke** [comment]
Database changes now in place on dev.
All events have hill_climb_ind set to 0. There are a small humber of hcs in the events table which will need correcting on live at some point.
There is a new table called hc_rider_history which has an identical structure to tt_rider_history. Currently empty.
Feel free to load some hcs into the events table whenever you are ready. No problem if not scored as we can delete / reload as needed. No rush on this - I have loads to do!

**Comment by rogernclarke** [comment]
Easiest part done - events page now has hill climb as an option in the distance filter dropdown.

One question though - I can't see where you've got the distance from for the few hill climbs we've loaded as they all seem to be 0 distance on CTT. All of them are quoted in miles - will that be the case for all of them? I know some in the past have traditionally been quoted in yards (presumably those < 1 mile).

**Comment by drnickw** [comment]
Some events have a distance but not many. Some have an average speed from
which you can derive the distance but sometimes its wrong (even negative in
one instance!).
These 3 I manually added the distance from the startsheet documents.

The YCF one is labelled a HC but was won at over 20 mph!

I`m wondering if we should have an HC season like the TT season (ie
excluding NDJ), perhaps last 10 days in August to first 10 in Nov. and only
score these events.
I guess within that if its labeled as HC I`ll have to accept it despite the
average speed. I think the ave speed for the gibbet is quite high and its
def a HC.
I`m inclined to stick to miles with a definition of up to 1/20 mile as in
the TTs.

**Comment by rogernclarke** [comment]
Fine with that. You'd have to include everything within the HC season marked as a hill climb - some years back the champs was on a very long gentle climb in the peak district and won by a well known time triallist not known for his climbing.

My only reservation about excluding out of hc season hill climbs is that those rides are entirely valid for qualification for national hc. I think they would be valid results as opposed to the out of season tts.

I've now got in place on dev the separate rankings screens for hc and tts - see top menu choices under Rankings - with a particularly unlikely hill climber in first place! I still need to style the menu but otherwise shouldn't need many further changes. I'm sure issues will crawl out of the woodwork once the data is in place.

**Comment by drnickw** [comment]
Yes, the problem with "out of season" HCs is that to score them properly
you need uptodate rankings and if we are only going to run the HC rankings
in season then i cant really validate the out of season rides points.
I`m happy to include everything which has "hill climb" as the label,
regardless of the average speed.
The problem again is the data is so bad, eg event 17787. mph is based
presumably on distance of "1 miles" as per ctt site but the actual distance
on strava is 0.57 miles.
I`m wondering if its even worth trying to get distances and speeds or if we
just use times alone. It would be nice to have the compete data set so i`ll
try, but i`ll have to see how much trawling around i have to do to get it
right.

**Comment by rogernclarke** [comment]
rider history table for hill climbs is now called tt_hc_rider_history.

Adding a row to this table updates the existing rider current table with the current hc ranking data to go with the current tt ranking data it already holds.

**Comment by drnickw** [comment]
I`ve finally got the program working to assign points to HCs. I`m now working on the rankings. I`ve got a run through from start to finish just iterating the ranking as it goes - ScoreddataHC-190802 and RankingsHC-SD-190802. Have a look and see if you think it looks right. Most people wont clock up enough rides to get out of provisional in 1 year so i think we need to change the scoring. I`m inclined to keep everyone in a single status which is effectively best but one of upto last 6 if possible.
I think we`ll need to have people "provisional" if they only have 1 ride however, or a flash in the pan single ride will be unbeatable.
For the reasons stated above and in the interests of building excitement over the 3 months I think it would be better to exclude last seasons rides from the ranking (although I`ll still use them for scoring the races), so riders will only start to build their current year score with their first races (mid August).
Not sure what this will mean for the website?
I could still score rides out of season for interest but as I`ll be using out of date rankings to do it and they wont be taken into consideration for the current rankings its probably better to not score them.
I can put the races and scores into the DB if you like, but there`ll be lots of missing fields until I do the full rankings run which is still probably 10 days away if I`m honest!! Would that be helpful or not?
Let me know what you think.

**Comment by rogernclarke** [comment]
Yes, please load up whatever data you can to dev as it is easier for me to look at it there - and it will undoubtedly throw up problems with the website display for me to get on with fixing.

I always thoughts the number of counting rides would have to differ for HCs. I had in mind that 3 would be a critical number - fairly sure that used to be the number of HC rides in the current and past year that you had to put on the National HC entry form in order to get a chance of a ride, and it seems more than reasonable to expect someone to get 3 HC rides in 12 months to qualify for a ranking.

We could see how many qualified riders that would give us. Not sure how you would then go about excluding best / worst rides as there wouldn't be many rides left! Maybe we shouldn't exclude any rides until someone becomes qualified and then we base their ranking on their best 3? I'm thinking an HC has less chance of a freak result - not many punctures and if you did you would DNF, and not sure how you would have a freak good result. Or maybe best 3 of up to 4 rides, best 4 of 5-6 rides, best 5 of 7+?

Would it be a good idea to seek the opinion of RG on this? We would have buy-in from someone from the HC scene if he agreed with the counting ride criteria.

I'm still not convinced that out of season HC rides are invalid. The rankings are no more out of date for scoring than they will be for the first in-season HCs, so the scores should be just as valid.

The question of restarting the rankings for the new season is a separate point. If you do that you are saying that the #1 at end of Nov last year suddenly has no status at all, and after week 1 of the season we'd have no qualified riders and the rankings wouldn't show anyone by default as they would all be provisional status. Presumably we could make a later call on this question as it wouldn't affect scoring each event.

**Comment by drnickw** [comment]
1 The question of restarting the rankings for the new season is a separate
point. If you do that you are saying that the #1
<https://github.com/rogernclarke/Rankings/issues/1> at end of Nov last year
suddenly has no status at all, and after week 1 of the season we'd have no
qualified riders and the rankings wouldn't show anyone by default as they
would all be provisional status. Presumably we could make a later call on
this question as it wouldn't affect scoring each event.

For me this is the main thing because i will affect how we rank. My concern
is the very real possibility that some big hitters from last year wont ride
this year at all but will remain right at the top of the rankings for most
of the season.

Andrew Feather 20872 1 2018102804 2.166667 1 F-16-15 y 2 3.666667
Calum Brown 10810 2 2019070501 3.666667 2 C-9-2 y 1 3.666667
Ed Laverack 10520 3 2018102804 4.5 2 P-4-4 y 3 4.5
Dan Evans 4965 4 2018102804 6.5 2 F-12-11 y 4 6.666667
Charlie Meredith 8016 5 2018093010 7 2 P-1-1 y 5 7
Adam Kenway 7899 6 2018102804 10.16667 3 F-10-9 y 6 10.16667
Kieran Savage -533 7 2017102801 11 3 P-1-1 n 7 11
Joseph Clark 11556 7 2019070501 15.16667 3 C-10-1 y 7 15.16667
Sam Lindsay 25743 8 2018102804 15.16667 3 F-12-12 y 10 16.83333
Richard Gildea 34 9 2019050512 15.66667 3 C-10-1 y 8 15.66667
Kieran Savage 15428 10 2018102804 15.66667 3 F-11-10 y 9 15.66667
Mike Morris 1846 12 2017102901 16 4 F-12-12 n 11 16
Aidan Quinn 17721 12 2018061720 16 4 P-1-1 n 11 16
James Lowden 2579 9 2017100808 17.5 3 P-2-2 n 9 17.5
James Hayward 6889 11 2018102804 17.6 4 C-7-6 y 11 17.6
George Kimber 13362 12 2018101406 18.2 4 C-7-5 y 12 18.2
Richard Bussell 4256 13 2018102804 19 4 P-1-1 y 13 19
Tom Bell 27092 14 2018102804 19.66667 4 P-3-3 y 14 19.66667
Max Stedman 11538 8 2017092310 20 3 P-1-1 n 8 20
Matthew Clements 13181 15 2018100711 21 4 P-1-1 y 15 21
Heres the top 20 and you can see 4 of them didnt ride in 2018. (date is
last event)
So rankings for HCs starting in September 2019 will have (if its anything
like last year) several people who are not competing this year.
As we will have a small number of rides for qualification its quite
possible these riders will remain qualified well into October.
This will of course affect the whole field, not just the top 20.

What I envisage is the site to run exactly as it has for TTs in the
background, with cats and predicitons based on the same old system which
from what I`ve seen of the HC data graphs for races still looks valid.
We wont need to change the counting rides for this as we predict and cat
irrespective of status.
However the rankings page alone, will reflect just this years progression.
I think thats what people will want to see.

I appreciate changing the system to exclude last year from the rankings
page is a big departure but I think HCs are much more seasonal and discrete
than TTs.
It would give us a real churn in the rankings for the first couple of weeks
which might be interesting and attract attention.

We could have 2 ranking pages, one for 12 months and one for this year but
I think that would be confusing and most people seem to look at the events
pages (which would still be driven by the classic TT system)
I`m happy to keep ALL the 2019 seasons results in the 2019 competition in
this situation.

Heres the figures for rides in 2018.

900 riders only 1
390 2
150 3
115 4
65 5
187 6 or more.
=====
1807

This is probably an over estimation as anyone riding early in the 2018
season as their last ride is likely to have 2017 results after that date
included too.
So half only ride 1.
Less than 28% ride 3 or more.
In a way the provisional to qualified point isnt so important as the cats
and predictions are all based on everyone anyway, its just a matter of who
we allow onto the rankings page.

By keeping the cats based on everyone over a 12 month period but allowing
the rankings table to build over the season gives us more inclusion and
excludes inactive riders from last year but also allows established
committed riders to know their value and be predicted appropriately and
come in in their rightful place when the do ride

2. Presumably we could make a later call on this question as it wouldn't
affect scoring each event.

Yes, I`ve already scored everything up to now, based on the rankings as of
old. I`ll try and get the fields I have, uploaded into the DB tonight.

**Comment by rogernclarke** [comment]
Ok, I think I've understood this but just to make sure...

You've got the underlying "pure" ranking that you use for predictions (it has a set of rules about which rides are included / excluded which don't take a rider's ranking status into consideration, so no advantage to frequent riders or penalty for provisional riders). This set of rules will be the same for tts and hcs

We then display the tt rankings using a different set of rules (which do take into account ranking status). We could term this to be a competition to distinguish it from the underlying ranking used for predictions.

For hcs we will also display a competition but the rules for this competition will differ from those for the tt competition - your proposal is to discount rides from the previous calendar year. We can tinker with the other rules for this competition up until the point we go live as it doesn't affect the scoring of rides.

A rider's category and predictions are based off the underlying ranking rather than the displayed competition standings. (I've just realised this isn't true for tts as category is based off the rules that incorporate frequent status, so I think there is a difference between the two there)

Is that an accurate summary?

If so then I guess we just need a proposal for the remaining rules of the hc competition, such as which rides get included / excluded.

I think the only thing I can see you losing by having a competition that restarts is having a visible list of how each rider has the category they have, which the tt rankings gives.

Maybe the concept of a competition as distinct from the ranking is something we can develop further at a later date.

**Comment by drnickw** [comment]
Yes, the ranking system underlying wont change and will be the same for TTs
and HCs. Currently this is the case for TTs,
The data I use to predict and score doesnt include frequent status and
there is no advantage to being qualified over predicted except you are
included in the rankings.
So, if you like, the fundamental ranking system is the same for both.
The TTs we add 2 rules - 1 for entry into the rankings which is 6 rides,
and 1 for frequent status. This is reflected in the rankings table only.
(currently i cat the TTs on the ranking table but maybe I should cat them
on the background data without the frequent status??)
The HCs we add new rules of our choosing. In view of the much smaller field
and frequency of rides i would propose qualified at 2 rides,  which still
means half the competitors this year wont qualify (unless the system
encourages them to ride 2 events). If we set it at 3 we are well under 1/3
qualifying.

The current rules for missing high and low seems to work well, which is 1-3
rides = average 4-8 rides = average minus max and min 9+ rides ave of
latest 8 minus max and min of latest 8.
Frequent is a special condition of TTs which is : best but one of any in
last 4 months if there are > 8 in last 4 months.

On the day we go live there will be about 240 riders who are ranked (ie
they`ve ridden this year) although almost non will have made the threshold
for the ranking table as they`ll only have 1 ride. This will be a problem
as the non provisional ranking table will be very small, but it should
populate quickly and only with current riders and should very soon become
appropriately ordered. Everyone however will know their cat (based on the
background ranking) as this will be shown against their name in the
events/riders pages.

I guess the problem with only needing 2 rides to qualify is the possibility
of someone taking the top spots with very low numbers of rides. Here`s the
list of riders with exemplary results (points <6)

Adam Kenway 30-Sep-18 Withington Wheelers 16328 02:14.2 3
Adam Kenway 07-Oct-18 sheffrec cc (male)(no lower age limit - closed road)
16500 01:16.6 4
Andrew Feather 29-Sep-18 mid-devon cc (event 2) (ride as one event or with
event 1 for combined K/QOM competition) 16779 04:44.0 1
Andrew Feather 13-Oct-18 Holme Valley Whs 17543 01:55.0 1
Andrew Feather 14-Oct-18 north lancashire road club  (NLTTA Championship)
16397 03:42.2 1
Andrew Feather 20-Oct-18 belper bc 16224 03:13.0 1
Andrew Feather 21-Oct-18 Matlock CC (£16 both events) 16225 03:16.5 1
Andrew Feather 28-Oct-18 RTTC National Hill Climb Championship  (DIW
Wheelers)(Entries close 09/10/18) 16471 02:18.8 2
Andrew Feather 30-Sep-17 mid-devon cc 16025 04:37.0 5
Calum Brown 15-Oct-17 cc sudbury (district championship) 16101 00:39.9 1
Calum Brown 07-Oct-18 sheffrec cc (male)(no lower age limit - closed road)
16500 01:14.5 1
Calum Brown 21-Oct-18 Matlock CC (£16 both events) 16226 01:36.8 1
Calum Brown 01-Jul-19 Matlock CC (Charity) 17874 00:46.1 1
Calum Brown 05-Jul-19 Sheffrec CC/Langsett Cycles RT (closed road - no
lower age limit) 18641 01:58.3 1
Calum Brown 14-Oct-18 tickhill velo cc (entries close 05/10/18) 16504
01:01.9 2
Calum Brown 28-Oct-18 RTTC National Hill Climb Championship  (DIW
Wheelers)(Entries close 09/10/18) 16471 02:20.5 4
Chris Crabtree 14-Oct-18 CC Sudbury (District Championship) 17451 00:40.1 1
Chris Newberry 30-Sep-17 sussex ca 16023 04:07.7 5
Dan Evans 13-Oct-18 Holme Valley Whs 17543 01:57.4 2
Dan Evans 14-Oct-18 Welsh CA (Welsh Championship) 17132 11:56.3 2
Dan Evans 23-Sep-18 Glossop Kinder Velo CC 16326 11:42.5 3
Dan Evans 23-Sep-17 exeter wheelers cc 15980 04:09.8 5
Dan Evans 07-Oct-18 lancashire road club 16394 02:23.5 5
David Huck 14-Oct-18 north lancashire road club  (NLTTA Championship) 16397
03:46.8 2
Ed Laverack 29-Sep-18 mid-devon cc (event 1) (ride as one event or with
event 2 for combined K/QOM competition) 16780 12:24.0 1
Ed Laverack 14-Oct-18 Welsh CA (Welsh Championship) 17132 11:55.9 1
James Hayward 21-Oct-18 Matlock CC (£16 both events) 16226 01:45.8 5
Joseph Clark 21-Oct-18 Matlock CC (£16 both events) 16225 03:21.9 3
Joseph Clark 20-Oct-18 belper bc 16224 03:19.1 4
Joseph Clark 07-Jul-17 sheffrec cc 15492 02:02.8 5
Jude Taylor 21-Oct-18 Huddersfield RC (Cheques to P Hurt) 17537 03:21.9 1
Jude Taylor 13-Oct-18 Holme Valley Whs 17543 01:59.3 3
Kieran Savage 13-Oct-18 Holme Valley Whs 17543 02:01.9 4
Kieran Savage 21-Oct-18 Huddersfield RC (Cheques to P Hurt) 17537 03:25.7 5
Leon Wright 21-Oct-18 Matlock CC (£16 both events) 16226 01:45.6 4
last digit is the points. (sorry about the poor spacing). I was encouraged
when i saw this that the system seems to be giving about the right number
of exemplary rides. (you dont want too many 1s or no one scoring below 25
say)
From this it looks like EL could have monopolised the top spot with his
average of 1 but infact he has an 8 too, which isnt shown. its possible we
may need to split close riders by freqency - not sure about this, but i
would think if there were no other rides for each rider, on the basis of
these results you might want AF and CB above EL. Its a fine point.

**Comment by drnickw** [comment]
This is easier to read as an email!

**Comment by rogernclarke** [comment]
Only 26% of non-lapsed tt riders have a C or F status, so requiring 3 rides to qualify gives a better % rate. I don't think 3 rides is too much to ask if you want to appear in the table with a complete status. Assuming that those with one or two rides would have provisional status and could be included with the provisional filter as per tt rankings?

Not yet convinced on using the current rules for missing high and low as I'm guessing the total ride count per rider is generally lower than for tts:
- 1-3 rides = average
- 4-8 rides = average minus max and min
- 9+ rides ave of latest 8 minus max and min of latest 8

Looking at your ride per rider data above, using 6+ as a frequent rider bonus point would act at a similar sort of % of riders as for tts. I think you meant for 9+ fastest 8 minus max and min not latest?

All rides are for current season so do we scrap the latest rides element of the rules as they are all within the timescale of the competition (almost all within last two months at the end of the competition)? We could then just say average of fastest 5 minus max and min. Gives a chance for riders doing 6+ to discard some duff rides as their frequent rider bonus.

If we go with qualification on 3 rides then it would also mean that discarding max and min rides could start on achieving qualification.

What do you think? Is there scope for trying out two models and seeing what difference there is?

**Comment by drnickw** [comment]
I dont have a strong view on when we bin min and max - what are you
proposing, the comments in purple are the status quo.
Not sure how much difference it will make dropping rides and as you say
you`re less likely to finish with a terrible time than to DNF.
9+ is latest - its only if the 9 are within 4 months that the frequent
rules come in and make it fastest, rather than latest.

Frequent as a concept I really like its just that over time I`ve realised
its not valid for ranking. It gives an unfair advantage.
However as we are moving the ranking subtly away towards a competition I
guess that doesnt matter and happy to include it from 6 rides if you would
like it in.

I`m more concerned about qualification. I agree if you only ride 2 events
you havent really entered into the spirit of it and so should you be in.
Its just that the ranking will look pretty thin for the first 2-3 weeks and
I`d like to engage with as many people as possible early on as its such a
short season and quite a number of the riders are not our regular audience.
I guess its a bit like the N+DCA. you can see your points building but
arent in the mix until you`ve got your qualification rides. I`d like for
somehow people who havent qualified to be able to see their progression to
qualification. Not sure about this but one idea could be to default the
rankings page to include provisional (but still with the option to exclude)
but grey them out till qualified. Perhaps you have a better idea as this
may give them too much credence.
I guess if we go with 3, people will still look at the predictions and cats
which wont be affected, and at their riders results. Its just the ranking
page which will be scant but I suspect if you arent a big hitter you
probably dont look at this (except with significant filtering).

**Comment by drnickw** [comment]
Currently working on the basis of 3 for qual, using all rides from calendar
year and frequent status from > 5 rides but only loosing min and max if >
3. so
1-2 both count provis
3 all count qual
4-5 cut to middle 2-3
6+ use best 3 minus very best
Is that how you see it?

**Comment by drnickw** [comment]
Currently working on the basis of 3 for qual, using all rides from calendar
year and frequent status from > 5 rides but only loosing min and max if >
3. so
1-2 both count provis                     ie A or A+B
3 all count qual                              ie ABC
4-5 cut to middle 2-3                      ie BC or BCD
6+ use best 3 minus very best  <<< should really say best 3 not including
the very best, so use average of 3 records whenever possible. ie BCD,
irrespective of how many there are.
Is that how you see it?

if races in best score order are ABCDEFG

**Comment by rogernclarke** [comment]
Yes, that's well expressed.

I've been contemplating how to show competition placings before any riders are qualified.

I don't want to change the rankings display component as it is shared code with the tt rankings. But I could try setting the tick in the Provisional box dependent on criteria if we can decide what the criteria would be.

Would it be for a duration such as before 1st October, or dependent on there being x number of qualified riders?

**Comment by rogernclarke** [comment]
Just a thought - if HCs now have a ranking and a competition, then what are we building up a history of in the hc_rider_history table?

For TTs all data in rider_history relates to the competition (very similar to the ranking but taking into account frequent status).

For HCs we can only get the category from the ranking (category applies to ranking instead of a competition). But then do we store historical placings week on week in the competition or historical placings in the underlying ranking? If it is the competition placings then they will reset each year. If it is the ranking data then it is storing information we currently have no plans to make available.

I'm guessing we go with the competition data and if we ever display that historical data in future we only show one season's competition at a time.

**Comment by drnickw** [comment]
That’s my intention, but if we want in future we will be able to show cat over time which will be a long term index, as well as competition for each individual year. Using the ranking data to create the rider Hx for HCs won’t make any sense as we want to see change week by week in the competition not the background rankings so all the positions (district, gender etc should be based on competition not ranking). 
I’m hoping I can get you a rider hx table this evening so we can have a look at it. It won’t be hard to change if we don’t like doing it like this.

otherwise

**Comment by rogernclarke** [comment]
I've put some code in place that checks if the rankings page is displaying hill climbs, and if it is checks to see how many riders are going to be displayed. If less than 20 then it will automatically display provisional riders included, otherwise it will go with the choice the user has made on whether to include provisionals or not (exclude by default). I'm not convinced I have the code correct so thought I should mention in case you see any strange behaviour once you have loaded the data!

**Comment by drnickw** [comment]
After the time it’s taken me to nearly populate these fields I think I can cut you some slack if things look a bit funny. I see one of August’s HCs has been cancelled. Maybe they’ll cancel the other and buy us a bit more time.

**Comment by drnickw** [comment]
I think you`re going to need a new ranking status. If a rider is a hill
climber and has rides from last year but not this year they are not lapsed
(or they wont have a cat) but neither are they provisional. Should I create
an H status for these and can you process them appropriately?
I see the first HC startsheet is already out!

**Comment by rogernclarke** [comment]
Does it need a status at all or could the status just be null? We have a null status for riders who have entered races but have no results, so wouldn't that be the same thing for an annual competition - could have entered something within the duration of the competition but has no results? That would work in the app with no changes I think and an H wouldn't cause us to want to display anything.

**Comment by drnickw** [comment]
Would a null status still show a category?

**Comment by rogernclarke** [comment]
I would guess so, but the existing null status riders have no category to show.

The field is now technically wrongly named as it is competition status. But category is related to the ranking instead so the two shouldn't by tied together.

If a null status doesn't show a category then I'll fix it so it does.

**Comment by drnickw** [comment]
Ive literally lobed some HC rankings in. I know of several issues still,
but wanted to give you a chance to see where we are at.
Clearly we wont be ready to predict this week but hopefully can go live
with HCs next wednesday if all goes well.
Let me know what you spot.
Nick

**Comment by rogernclarke** [comment]
Thanks. Lots of issues for me to work on - almost don't know where to start!

I've removed what I put in last week to include provisional riders on the rankings list if less than 20 qualified a it was confusing things. Will look at reinstating once everything else is working.

One issue that I think is yours - lots of provisional ranked riders with a rank of zero. Noticed this is happening again on live - 19 of them.

**Comment by drnickw** [comment]
Yes, I saw the zero scores on dev. I hadn’t seen them on live. I’ll need to look at that again.

The other things I am aware of from my point of view are too many events where we haven’t scored cos the data’s dodgy, I can fine tune this - sometimes the top half of the field correlates well but then the data becomes poor. Do you think people would be upset if I score only the top half of the event!

 Also some errors with club and age cat on negative riders. There are also some rider hx lines with a blank cat which I’m unsure about although I think all the current lines are correct. Plenty to go at in the next few days.

**Comment by rogernclarke** [comment]
I'll fire these off as I find them...

Next one for you - we've talked in the past about what the ride_distance column on the rides table should hold. It was intended for duration events. Can we set it to null for HCs and distance events?

I'm fixing my code up to handle it for now. Thought I'd start with the basic display items before hitting the rankings!

**Comment by rogernclarke** [comment]
Event id 17781 quite a few results are missing including first place!

**Comment by rogernclarke** [comment]
category_on_day is null for all hill climb rides. I think it should be populated with the rider's hill climb category on the day

**Comment by rogernclarke** [comment]
counting_ride_ind needs to be populated on each hc ride record so I know which ones count towards the ranking (competition).

As an added complication for HCs, I think we need to know the counting rides for each year's competition. I will need to add a Year dropdown box on the HC ranking screen so people can choose which year's competition to view. We'll therefore need to know which rides counted in each of the years, so indicate at the end of each season which rides counted for that year.

The ranking screen for tts currently displays all events in the last 12 months and then shows which ones are counting. For HCs this needs to be all HCs in the selected calendar year and then show which ones are counting.

**Comment by rogernclarke** [comment]
I can't tie up a rider's ranking score (from hc_rider_history - current record) with their ride scores. E.g. rider_id 5447 has one ride scoring 12 but a ranking score of 4? This rider's current category on hc_rider_history is also blank.

**Comment by rogernclarke** [comment]
event_id 18391 has no positions for the results but does have times!

**Comment by drnickw** [comment]
Keep them coming. I’ll get onto them asap, hopefully some this evening, depending on how the weekly run goes and whether we go to the pub or not.
I’ve found out why the live dB ranks have some zero provisionals - felt that was the priority. Should correct with tonight’s run.
If it takes longer to get the HCs in at least the ranking page will be well populated from the start!
As an aside, and I’m just being a bit lazy here, to check, do I currently set any fields to NULL? NULL isn’t a valid value for a variable in R, it clears the variable away, so I can set to NA or 0 or “ “ If things are coming through as NULL the dB must be translating them.

**Comment by drnickw** [comment]
I hadnt updated the rides table for a while. I think I may have solved a
lot of the problems by updating that.
Check 18391 and 17781. The 15316 has some dodgy club names which I`m onto.
5447 is right, the ride score had changed.
I still need to sort out the blank cats.
Think you`ve got CRindex now?

So its blank cats, NULLs, zero provisionals and previous years counting
rides, assuming the above are right now.
Did I miss anything else?
Sorry, I should have reloaded the rides - as I said I just lobbed a RHx
table in!

**Comment by rogernclarke** [comment]
Looks much better - but looks like the times have gone back into mm:ss.s instead of hh:mm:ss.s?

You've lost me with CRindex?

I think there must be lots of columns set to null - predicted distance? Maybe try just not setting the distance column? This isn't top priority though as I'm working round it.

Think those are the only things I've passed your way. I've spotted one or two other things I need to fix. Easier to see what is going on now.

Won't be able to do any more on it until the weekend now.

**Comment by rogernclarke** [comment]
I think I have done as much as I can now. I have two major issues to sort out but we can potentially go live with it if we are happy with two workarounds...

Firstly, I've got a problem with adding the filter for HC rankings to filter by year. I could add it but it appears briefly for tt rankings as well before getting hidden, not sure why. The problem if I did add it is that the current rankings display runs off rider_current, which is fine for the current year's competition, but doesn't hold the data for previous years. So I'd have to amend the code to look for the last rider_history records of the relevant year. I can do it but not for going live this week. We can go live with it as is, hard coded just to display the 2019 competition standings and then I could do a subsequent release to add options to display previous years. I think that would be fine rather than delaying going live.

Secondly, I was trying to force the provisional riders to be displayed if less than 20 riders are qualified. I've changed my mind that this would be a good idea as it doesn't then allow a user to hide provisional riders if less than 20 are qualified. That would also apply when the user has selected filters to trim down the display to district etc. it would be ok if it defaulted to show provisionals and allowed you to untick it but I don't think I can get it to do that. Again, i think better to go live like it is and look for options later if needbe.

How are you getting on with the other bits and pieces?

I would be looking towards Thursday evening for putting it live I think, if we are ready this week - better not to mix it up with the weekly run.

**Comment by drnickw** [comment]
Hi Roger,
Thanks for that, I think the filter by year is an enhancement at present so
dont see a problem with missing it just now.
The provis tick box was a bit of a fudge anyway, the only problem being
currently there are only 3 qualified riders and they are not big players
(cat a13,c20 and d8)
Its going to look at first glance a bit weak. I know our FAQs will explain
what we are doing but i think first impressions will be poor and it seems
the majority of people still havent read the current faqs or really
understand the system.
I`ve been beavering away, trying to get my head around scoring and ranking
from 2 different systems and making all the ranks and cats line up. Its
been a bit frustrating but i`m hoping for a breakthrough soon.
I wonder if we should give it another week or two, as I dont want to rush
it and it be a bodge, and with a few more races under our belts there will
be more to see ranking wise I hope.
I`m certainly not ready now but if all the cards align I might be stable
for Thursday.
Nick

**Comment by drnickw** [comment]
I`ve fixed what im aware of, except the NULLs which i think you have a
handle on. I`ve put the status to blank if they are not actice this year
but still have events within 12 months so not lapsed. however it still
seems to be showing as provisional on the rankings page.
I screwed up and put the HC rider hx into the tt rider hx table by mistake
the first time. I dont know if you can recover this or just want to delete
and I`ll put the tt rider hx back - or you can copy it from live.
When I go onto the rankings page initially it doesnt give me a page puck
list at the bottom, but it comes up later if i change the number of riders
on screen.

See Ian Mountjoy 11264. says hes cat D1 provis for HC but his latest HC
rider hx line is D3. (im only creating rider hx lines for weeks there are
any HC results - is this a problem?) Not sure where the D1 is coming from -
may be related to my corruption of TT rider hx?

**Comment by rogernclarke** [comment]
I'm 99% certain it is the loading of history records that has caused the issues you are seeing. Inserting records to either rider_history table pushes through the data to rider_current. So the current tt rankings data is showing the hc history record data you loaded and the hc data is old.

If you do a load to dev tonight of this week's tt and hc history records to the right tables then I'm hoping it will sort itself out. I should then be able to see what if anything else needs looking at.

I'm then inclined to take a fresh copy from live to dev and do a dry run for how we are going to put hcs live as there are several steps and I don't want to screw it up on live. Will give us a chance to see which steps can be done whilst leaving the live website available.

**Comment by rogernclarke** [comment]
I think I've found an issue for you to look at - event_id 19033, winner is shown as cat E3. Looks like he has always been an HC A cat rider.

**Comment by rogernclarke** [comment]
And another query - why are there too few ranked riders to award ranking points on event_id 17874?

**Comment by rogernclarke** [comment]
Here is a list of the outstanding fixes I need to do (I've fixed some others already):

- Awards on rider screen don't show decimal place for hill climbs
- Hill
- Climb distances of zero currently just show as a hyphen. On rider awards list it isn't always obvious which events were HCs
- HC team aggregate result isn't adding up correctly - rounds seconds on each team member before adding up

In addition, when we add in predictions need to check handicap award results.

**Comment by drnickw** [comment]
1. rider showing E7 winning. I did have some problems with adjusting the
category based on the number of riders eligible for categorisation which
was why the rhx was up the creek for a while. I havent re run the whole
system from scratch yet, so the wrong cats are still on some of the ride
results and clearly this is where you are picking this data up from.  I`d
forgotten you`d not just get the cat from the rhx.
I need to re run the points and it will correct all the cats, and then I`ll
need to re load rides and rider hx. is it ok for me to just truncate these
tables and replace them when im done?
2. one of the new bits of verification is to see if the points on the time
vs rank graph have any correlation. sometimes its so weak that although
there may be lots of points I cant draw a valid progressive line through
them. In this situation i`ve chosen not to score. I`ve not done this for
TTs which probably goes some way to explaining all the draws predicted in
the recent 10 as the graph is too flat. I alluded to this in a previous
message, as it would be possible to score the parts of the results which do
correlate but then some people would get a result and others in the same
race would not. Or I could just invent a gradient and score on the basis of
this. Not quite decided what to do yet so left them out. It seems to apply
a lot more to HCs as the data isnt as good or as numerous, and especially
with the early season/2017 results. Ideally I need to resolve this before
we go live to prevent retrospective changes but also its not the end of the
world if something doesnt get scored but from my point of view better that
than poor quality results (others may take the opposite view). I guess at
the least the message needs to be changed to "unable to deduce a valid
points score from the data available" or some such.

**Comment by rogernclarke** [comment]
Fine to delete the rides and hc rider history records. You can truncate hc rider history if you are reloading all the records but the rides table is shared with tt rides so I guess you only want to delete rides for hc events?
If you want to let me know what the new message should be for explaining why any event isn't scored then I'll update the message.

**Comment by drnickw** [comment]
would be "not enough reliable data to produce valid points scores" - this
would cover all eventualities.
I`ll consider a generic graph override in future to score these events. I
could do it in a similar way to handling the top end if there is any area
of useful correlation to pick up on.

**Comment by rogernclarke** [comment]
Text now updated

**Comment by drnickw** [comment]
I`ve now updated the tables - the cats should all be right and the points
are adjusted so its harder to get a 1. should limit the likelihood of draws.
I think thats me for HC except for  FAQ, predictions and previous years
counting_ride_index - in that order of priority. Clearly need FAQ before
going live. Do you want to wait for predictions or not?

**Comment by drnickw** [comment]
Theres a lot of needing to reload the HC data when using the back button
and advancing through the website currently. Can it be set up so once
you`ve chosen HC it stays on HC till you choose TT?
For example if you go onto HC rankings and click on the first rider it
takes you to his TT rides page and if you change this to HC page and go
onto his first HC and go back it goes back to the TT page again.

**Comment by drnickw** [comment]
*I`ve put a first draft FAQ in a folder on the shared dropbox account.*
*I`m sure there  will be other things to mention, and i cant remeber how
many TT events there are in a year so feel free to complete the "nnn" if
you can.*
*happy for any thoughts or additions.*

**Comment by rogernclarke** [comment]
FAQs look good. I've made a few non-material tweaks and added in the TT count (850).

I've just got two trivial issues left, but I'd go live with it as is.

Going straight to HCs on rider page when hitting the back button isn't simple. Clicking on TT / HC / Awards doesn't actually load any data - it just hides some html and shows some other html. The starting position when you load the rider page (and therefore the same when you go back to the page) is to show the tt results. To make it work differently we'd have to store something in the user's browser (in a cookie) to indicate they are viewing hc or tt and then query that on page load to display the right things. Not something I know how to do at present...

**Comment by rogernclarke** [comment]
A very minor potential issue with hc rankings.
Look on the hc rankings screen at the counting rides for the rides in first and fourth places. Both have their top rides tied on points with their second ride, but one has the most recent ride discounted and one has the older ride discounted. Obviously it looks better if the top listed ride is the discounted one. Not sure if you are going to say that the scores aren't identical due to decimal places... but I think we had this same question on tts and it was resolved somehow.

**Comment by rogernclarke** [comment]
I'm going to close this as we have put hill climbs live. Will open a new issue for displaying previous year's competition results
